src/pstl_offload.cpp

```diff
@@ -113,43 +120,236 @@ __original_msize(void* __user_ptr)
     return __orig_msize(__user_ptr);
 }
 
+static auto
+__get_original_realloc()
+{
+    using __realloc_func_type = void* (*)(void*, std::size_t);
+
+    static __realloc_func_type __orig_realloc = __realloc_func_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+#elif _WIN64
+
+static __free_func_type __original_free_ptr = free;
+
+#endif
+
+inline bool
+__is_ptr_page_aligned(void* __p)
+{
+    // using that __get_memory_page_size() returns only power of 2 values
+    return ((std::uintptr_t)__p & (__get_memory_page_size() - 1)) == 0;
+}
+
+struct __hash_aligned_ptr
+{
+    std::uintptr_t operator()(void* __p) const
+    {
+        // We know that addresses are at least page-aligned, so, expecting page at least
+        // 4K-aligned, drop 11 right bits that are zeros, and treat rest as a pointer,
+        // hoping that an underlying Standard Library support this well.
+        constexpr unsigned __ptr_shift = 11;
+        // current page size is at least 4K
+        assert(__get_memory_page_size() >= (1 << __ptr_shift));
+        return std::hash<void*>()((void*)((std::uintptr_t)__p >> __ptr_shift));
+    }
+};
+
+template <class T>
+struct __orig_free_allocator
+{
+    using value_type = T;
+
+    __orig_free_allocator() = default;
+
+    template <class U>
+    constexpr __orig_free_allocator(const __orig_free_allocator<U>&) noexcept {}
+
+    T* allocate(std::size_t __n)
+    {
+        if (T *ptr = static_cast<T*>(std::malloc(__n * sizeof(T))))
+        {
+            return ptr;
+        }
+        throw std::bad_alloc();
```

Comment:
May be it would be more readable if we revert the logic?
if (!ptr) {
    throw bad_alloc{};
}
return ptr;

As far as I know the original code can result in warnings on some system because the else branch don't have a return statement.

Reply:
Done.

Very strange systems, I would say. In which case they expect `throw` not throwing?